src/clientiface.cpp

```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
-			if (block == NULL || block_not_found) {
-				if (emerge->enqueueBlockEmerge(peer_id, p, generate)) {
-					if (nearest_emerged_d == -1)
-						nearest_emerged_d = d;
-				} else {
-					if (nearest_emergefull_d == -1)
-						nearest_emergefull_d = d;
-					goto queue_full_break;
+			if(block == NULL || surely_not_found_on_disk || emerge_required)
+			{
+				bool allow_generate = true;
+				if (!emerge->enqueueBlockEmerge(peer_id, wms.p, allow_generate)) {
+					// EmergeThread's queue is full; maybe it's not full on the
+					// next time this is called.
 				}
 
-				// get next one.
+				// This block is not available now; hopefully it appears on some
+				// next call to this function.
 				continue;
 			}
 
-			if (nearest_sent_d == -1)
-				nearest_sent_d = d;
-
-			/*
-				Add block to send queue
-			*/
-			PrioritySortedBlockTransfer q((float)dist, p, peer_id);
-
-			dest.push_back(q);
-
-			num_blocks_selected += 1;
+			// The block is available
+			return wms;
 		}
-	}
-queue_full_break:
-
-	// If nothing was found for sending and nothing was queued for
-	// emerging, continue next time browsing from here
-	if (nearest_emerged_d != -1) {
-		new_nearest_unsent_d = nearest_emerged_d;
-	} else if (nearest_emergefull_d != -1) {
-		new_nearest_unsent_d = nearest_emergefull_d;
-	} else {
-		if (d > full_d_max) {
-			new_nearest_unsent_d = 0;
-			m_nothing_to_send_pause_timer = 2.0f;
-		} else {
-			if (nearest_sent_d != -1)
-				new_nearest_unsent_d = nearest_sent_d;
-			else
-				new_nearest_unsent_d = d;
+		if (wms.type == WMST_FARBLOCK) {
+			// Not supported; always discard
+			continue;
 		}
 	}
 
-	if (new_nearest_unsent_d != -1)
-		m_nearest_unsent_d = new_nearest_unsent_d;
+	return WantedMapSend();
+}
+
+void RemoteClient::cycleAutosendAlgorithm(float dtime)
+{
+	m_time_from_building += dtime;
+	m_autosend.cycle(dtime, m_env);
 }
 
-void RemoteClient::GotBlock(v3s16 p)
+void RemoteClient::GotBlock(const WantedMapSend &wms)
 {
-	if (m_blocks_sending.find(p) != m_blocks_sending.end()) {
-		m_blocks_sending.erase(p);
-		// only add to sent blocks if it actually was sending
-		// (it might have been modified since)
-		m_blocks_sent.insert(p);
+	if (m_blocks_sending.find(wms) != m_blocks_sending.end()) {
+		if (wms.type == WMST_MAPBLOCK)
+			m_mapblocks_sent[wms.p] = m_blocks_sending[wms];
+		else if (wms.type == WMST_MAPBLOCK)
+			m_farblocks_sent[wms.p] = m_blocks_sending[wms];
+		m_blocks_sending.erase(wms);
 	} else {
 		m_excess_gotblocks++;
 	}
 }
 
-void RemoteClient::SentBlock(v3s16 p)
+void RemoteClient::SendingBlock(const WantedMapSend &wms)
 {
-	if (m_blocks_sending.find(p) == m_blocks_sending.end())
-		m_blocks_sending[p] = 0.0f;
-	else
-		infostream<<"RemoteClient::SentBlock(): Sent block"
-				" already in m_blocks_sending"<<std::endl;
+	assert(!m_blocks_sending.count(wms) ||
+			m_blocks_updated_while_sending.count(wms) &&
+			"Block shouldn't be re-sent if it is being sent and was not"
+			" updated while being sent");
```

Comment:
should be a `FATAL_ERROR_IF` or print to `warningstream` depending on conditions